libs/cln-version-manager/clnvm/cln_version_manager.py

```diff
@@ -0,0 +1,263 @@
+    ):
+        self._cln_path: Path = _get_cln_version_path(cln_path)
+        if cln_versions is not None:
+            self._cln_versions = cln_versions
+        else:
+            self._cln_versions = CLN_VERSIONS
+
+    def get_versions(self) -> List[VersionDescriptor]:
+        """
+        Retrieves the list of Core Lightning versions
+        """
+        return self._cln_versions
+
+    def is_available(self, cln_version: VersionDescriptor) -> bool:
+        target_path = self.get_target_path(cln_version)
+        return os.path.exists(target_path)
+
+    def get_all(self, force: bool = False) -> Dict[str, NodeVersion]:
+        """
+        Downloads all versions of Core Lightning
+
+        Args:
+            force (False): If force is True the data will be overriden
+        """
+        versions = list(self.get_versions())
+
+        def do_download(
+            version: VersionDescriptor,
+        ) -> Optional[Tuple[str, NodeVersion]]:
+            try:
+                return version.tag, self.get(version, force)
+            except Exception:
+                logger.exception(
+                    "Failed to download %s. This version will be ignored", version.tag
+                )
+                return None
+
+        def do() -> List[Tuple[str, NodeVersion]]:
+            with Pool(10) as p:
+                data = p.map(do_download, versions)
+
+            return [t for t in data if t is not None]
+
+        return dict(do())
+
+    def get_target_path(self, cln_version: VersionDescriptor) -> Path:
+        """Computes the path corresponding to which a cln version should be downloaded"""
+        return Path(self._cln_path) / cln_version.checksum / cln_version.tag
+
+    def get_descriptor_from_tag(self, tag: str) -> VersionDescriptor:
+        cln_dict = {d.tag: d for d in CLN_VERSIONS}
+        descriptor = cln_dict.get(tag, None)
+
+        if descriptor is None:
+            raise ValueError(f"Failed to find version {tag}")
+
+        return descriptor
+
+    def latest(self) -> NodeVersion:
+        vs = [d.tag for d in self.get_versions()]
+        latest = max(vs)
+        descriptor = self.get_descriptor_from_tag(latest)
+        return self.get(descriptor)
+
+    def get(self, cln_version: VersionDescriptor, force: bool = False) -> NodeVersion:
+        """
+        Ensures the provided version exists.
+        It returns the path to the corresponding binary
+        """
+        target_path = self.get_target_path(cln_version)
+
+        if not os.path.exists(target_path):
+            self._download(cln_version, target_path)
+        elif force:
+            shutil.rmtree(target_path)
+            self._download(cln_version, target_path)
+        else:
+            # Files are already downloaded
+            # We don't do anything
+            pass
+
+        return NodeVersion(
+            name=cln_version.tag,
+            path=target_path / "usr" / "local" / "bin" / "lightningd",
```

Comment:
Nit: Path components can be put into a list variable for re-use later in this file:
https://github.com/Blockstream/greenlight/pull/380/files#diff-d7ea266fb9501943d307089bf8a6c69645c28b8b750303519109e7c331f69d0bR247

It could also be imported in cli.py so we can get its length instead of the hardcoded `4` in `-4`:
https://github.com/Blockstream/greenlight/pull/380/files#diff-052bff27cf16d61abb6af6d16098e6e0287aaceeae377738de20be4a3eef9d38R71

Reply:
I've changed the approach here a bit.
I decided to handle paths a bit more explicitly.

I've introduced new fields in `NodeVersion` which can be used to query the path.
Users of our library should never use `[:-4]` selectors to get the path they need.

PS: Currently, I still use it in `gl-testing`. However, note that `gl-testing` isn't using the managed versions yet.